### DISCUSSION TITLE
Back-end database access

### DISCUSSION THREAD

**Discussion by Batimius** [pocketbase/pocketbase#3829] [Q&A] [answered]
Hello there. I am new to using any type of database, so please forgive me for my lack of knowledge. In short, I am creating a Next.js app, and every interaction with the database is done in the back-end (no direct requests to the database itself). That said, I was curious about one thing.

Would it be better to create, read, and update entries using an admin account that I log in once on runtime and then utilize the token of, or would it be better to auth the user whenever they sign in and hold a copy of their token to then utilize for every request they make?

I am unsure how performant the second one is, I just thought about it from the perspective that due to the limited permissions a user has, any accidental requests (in case I ever miss something) would get rejected, meanwhile with an admin account, they wouldn't. Either way, I am not familiar with databases (I've only been using JSON files up until now), so any suggestions are appreciated. Thank you for your time and help, they are truly appreciated.

**Comment by whmeitzler** [comment]
Pocketbase acts as a backend already- it's a database (SQLite) with a data model (collections and auth) and an API (http/json).

Next exposes an API, but less so a data model and no database.

If you're just starting out, it's a good idea to get comfortable with what these programs do. If you've already explored next with a file-based backend (which is perfectly fine for learning), I suggest letting Pocketbase act as your entire backend. Get accustomed to what it does. Explore the code. Try extending it with Go or JS. Let it be a grok'able backend you can explore.

Looks like auth, security, and permissions are going to be the next domain to explore. Someone, somewhere, sometime, has to make the decision as to whether a request can be fulfilled. Focus on understanding that decision process first, before worrying about performance.

Build something cool from there 😎

**Reply by Batimius** [reply, marked as answer]
Hello there. Thank you for the reply! After experimenting a little, I ended up coming up with a custom solution that still utilizes PocketBase, but purely as a database.

For my application, once a user signs in / up (using OAuth), I want to keep them signed in, which means creating custom sessions (session token + persistent token). With the current way PocketBase is set up, authenticating based on said sessions would be impossible since the integrated OAuth authentication system uses a popup instead from what I understand (I can't just pass up an access token and call it a day).

What I ended up doing was creating a separate collection that holds the user information (username, openid, and refresh token), a collection for the OAuth tokens (since they are not persistent and need to be refreshed once in a while), and lastly, one for my own session tokens. These are linked to each other (session tokens point to a user, OAuth tokens point to a user, user points to a single OAuth token) and use one another when needed.

Unfortunately PocketBase doesn't really allow me to create "virtual sessions" with just an OAuth token, but I'll still be utilizing it as a database regardless. Thank you for the help!

**Reply by ganigeorgiev** [reply]
> For my application, once a user signs in / up (using OAuth), I want to keep them signed in, which means creating custom sessions (session token + persistent token).  With the current way PocketBase is set up, authenticating based on said sessions would be impossible since the integrated OAuth authentication system uses a popup instead from what I understand (I can't just pass up an access token and call it a day).

> Unfortunately PocketBase doesn't really allow me to create "virtual sessions" with just an OAuth token

I'm not sure that I understand what do you mean by this, but both claims seems incorrect to me.

I'll not pursue the matter further since you already found a solution to your problem, but to avoid eventual confusion if someone stumble here - note that once OAuth2 authenticated PocketBase returns both regular PocketBase auth token and the tokens related to the OAuth2 session (_the latter are returned in case you want to perform further API requests against the OAuth2 provider apis_).

> but I'll still be utilizing it as a database regardless

As mentioned below, if you are not going to make use of the PocketBase db abstraction and web APIs, you may not need PocketBase and may be better with just using SQLite directly.

**Reply by 2mac** [reply]
Is there documentation for managing these tokens? I see how to retrieve the token from the `authStore` but I haven't found where I could set it again later, say from a cookie.

**Reply by ganigeorgiev** [reply]
@2mac PocketBase is fully stateless and there are no sessions in the traditional sense (even the issued tokens are not stored in the database). A request is considered authenticated if it has a valid `Authorization: TOKEN` header with it.

The SDKs `authStore` is a helper that just holds the token locally and for convenience add the specific header with each following request.

If you want to load the store from a cookie, you can make use of the available helpers `loadFromCookie(...)` and `exportToCookie(...)` that are commonly used in JS SSR scenearios (see https://github.com/pocketbase/js-sdk#ssr-integration).

For all available authStore methods for the JS SDK you can also check https://github.com/pocketbase/js-sdk#common-auth-store-fields-and-methods.

With that said, if you are trying to emulate a session-like behavior with cookies, you can check the JS hooks and Go examples in https://github.com/pocketbase/pocketbase/discussions/3447 or https://github.com/pocketbase/pocketbase/discussions/989, but PocketBase wasn't really designed to work this way and the dev experience may be subpar.

Please note that no one is selling anything here and there is no point sugarcoating the current implementation shorcomings. PocketBase is just a tool that I created primarily for my own specific use case (see [Presentator#183](https://github.com/presentator/presentator/issues/183)). Other use cases that deviates too much from this and what already is implemented may not be suitable for PocketBase (at least not in its current stage) and you may be better with using plain SQLite or other BaaS-like solutions (just to mention some - Supabase, Nhost, Appwrite, EdgeDB, SurrealDB, Strapi, Directus, Payload, etc.)

**Comment by 2mac** [comment]
I'm very familiar with databases but new to PocketBase. One concern I have along these lines is requiring frontend JS to even log in. I like to make my webapps as compatible as possible, with minimal requirements for client-side JS. How is one supposed to handle authentication in a case like this? For querying the database, it's simple enough to just have my own web API with a PocketBase session to query, but auth seems to be tightly integrated and not designed for compatibility.

**Reply by ganigeorgiev** [reply]
I'm not sure I follow.

Could you elaborate what do you mean by _"I like to make my webapps as compatible as possible"_?

>  For querying the database, it's simple enough to just have my own web API with a PocketBase session to query

Why do you need your own API?

What is the point of using PocketBase if you are not going to make use of its available [web APIs](https://pocketbase.io/docs/api-records/) and db abstraction?

PocketBase is designed to be used and interact directly. That's why it has a builtin web server in the first place. There will be very little to none benefit of wrapping it with another web API.

If you just want to use SQLite and already have your own web API, then you may not need PocketBase.

**Comment by 2mac** [comment]
> Could you elaborate what do you mean by "I like to make my webapps as compatible as possible"?

By this I mean I prefer for my app to work without any client-side JS at all. [Not everyone has JS](https://www.kryogenix.org/code/browser/everyonehasjs.html), and even for those who do, JS is far more likely to experience in-session errors (e.g. caused by connection drop-outs or issues with state) than HTML and CSS which don't fail at all once they load.

Obviously some features will require JS, and I'm not opposed to using JS to implement them, but basic functions like logging in and submitting forms should not require JS.

> Why do you need your own API?

Because PocketBase doesn't understand basic HTTP form submission and instead relies on API calls.

>What is the point of using PocketBase if you are not going to make use of its available [web APIs](https://pocketbase.io/docs/api-records/) and db abstraction?

I do make use of its web APIs and abstraction *from the server* whilst processing web requests. Basically the only thing I can't do this way is real-time updates.

> If you just want to use SQLite and already have your own web API, then you may not need PocketBase.

PocketBase still offers some advantages:

- Collections can be set up in a web UI
- Changes to the database are automatically exported to a migration script for later deployment

It is becoming more apparent to me, though, that PocketBase is opinionated about app architecture.

**Comment by whmeitzler** [comment]
> Because PocketBase doesn't understand basic HTTP form submission and instead relies on API calls.
I think you're seeing the focus on JS/Dart frontend libs and missing out on the fact that those libs interface with a Pocketbase backend entirely through HTTP. In my experience, Pocketbase's API tier is exciting because it's so boring- very straightforward REST, and even realtime is SSE instead of Websockets/gRPC/whatever is in vogue these days. 

I suggest poking around the "api preview" pages in the web UI and testing out some flows with cURL or similar. Auth is best explained there. There is also the [docs](https://pocketbase.io/docs/authentication), but that tends to explain things in terms of frameworks. 

If the stock APIs suit your needs, you can serve static pages using `pb_public`, or use Pocketbase as a framework for server -rendered pages like it's 1999.

@ganigeorgiev, if I may make a suggestion, your docs explain how to talk to Pocketbase in terms of your JS and Dart libraries. If you added another tab for cURL/REST/HTTP, that laid out the headers and endpoints, it may make it easier for people to use (or discover how they can use) Pocketbase with simple REST clients. I know I've wished for a cURL tab in the API preview that I can copy/paste for local development at times.

**Comment by 2mac** [comment]
> I think you're seeing the focus on JS/Dart frontend libs and missing out on the fact that those libs interface with a Pocketbase backend entirely through HTTP.

I think you're missing the part where I don't want to rely on client-side JS for auth and data loading. I realize that all the traffic with PocketBase is happening over REST/HTTP, but the fact is that everything is designed (or at least documented) under the assumption you will connect to PocketBase from the browser using JS and that you would never want to handle auth/database transactions on the server.

It really doesn't matter whether PocketBase uses HTTP or some proprietary socket protocol for its transactions. If it requires JS on the browser, then it fails to achieve my goal.
